### PR TITLE
Add scalable gossip library

### PR DESCRIPTION
**Pull request by aeyakovenko**
Implements a scalable gossip protocol.

### CODE REVIEW

**Comment by garious** [comment]
What I see here is 3 independent uses of crdts all thrown into one structure awkwardly named `Crdt` 10 days after we merged a PR to stop using that term as struct name, https://github.com/solana-labs/solana/pull/1450.

**Comment by aeyakovenko** [comment]
@garious its just not type agnostic.  This actually implements a crdt with a merge strategy for all the different Value types.

**Comment by aeyakovenko** [comment]
@garious what do you think about this version?  i got rid of the `crdt` names, and explicitly documented what this replicates

File: src/lib.rs [hunk 1]
```diff
@@ -86,17 +94,20 @@ pub mod window;
...
 extern crate bs58;
+extern crate bv;
 extern crate byteorder;
 extern crate bytes;
 extern crate chrono;
 extern crate clap;
 extern crate dirs;
 extern crate elf;
 extern crate generic_array;
+<<<<<<< HEAD
```
Comment (garious) [inline src/lib.rs]: Looks like it's about time to turn CI back on.

**Comment by aeyakovenko** [comment]
@mvines this seems to work!  i would like to merge before i get another big set of conflicts.

**Comment by mvines** [comment]
@rob-solana / @carllin - please review for content too.  You guys are more familiar with this part of the code than I

File: src/crds_gossip_pull.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
+//! Crds Gossip Pull overlay
+//! This module implements the anti-entropy protocol for the newwork.
```
Comment (mvines) [inline src/crds_gossip_pull.rs]: "newwork"

File: src/crds_gossip_pull.rs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+//!
+//! The basic strategy is as follows:
+//! 1. Construct a bloom filter of the local data set
+//! 2. Randomly ask a node on the network for data that is is not contained in the bloom filter.
```
Comment (mvines) [inline src/crds_gossip_pull.rs]: "is is"

File: src/crds_gossip_pull.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+//!
+//! Bloom filters have a false positive rate.  Each requests uses a differnet bloom filter
```
Comment (mvines) [inline src/crds_gossip_pull.rs]: "differnet" -- heh, we gotta get you an editor with spelcheck!

File: src/crds_gossip_push.rs [hunk 5]
```diff
@@ -0,0 +1,453 @@
+//! Crds Gossip Push overlay
+//! This module is used to propagate recently created CrdsValues across the network
+//! Eager push strategy is based on Plumtree
+//! http://asc.di.fct.unl.pt/~jleitao/pdf/srds07-leitao.pdf
+//!
+//! Main differences are:
+//! 1. There is no `max hop`.  Messages are signed with a local wallclock.  If they are outside of
+//!    the local nodes wallclock window they are droped silently.
```
Comment (mvines) [inline src/crds_gossip_push.rs]: "droped"

**Pull request merged**